xrtl/base/threading/message_loop.h

```diff
@@ -0,0 +1,324 @@
+//  loop->MarshalAsync(&task_list, []() {
+//    // Runs asynchronously at some point in the future on the loop thread.
+//  });
+//  OnButtonPressOrWhatever([]() {
+//    loop->Exit();  // Start shutting down the loop.
+//  });
+//  Thread::Wait(loop);  // Wait until the loop exits.
+class MessageLoop : public WaitHandle {
+ public:
+  class Task;
+  class TaskList;
+
+  // Creates a new message loop thread.
+  static ref_ptr<MessageLoop> Create();
+
+  virtual ~MessageLoop();
+
+  // Returns true if the currently executing code is running on the loop thread.
+  virtual bool is_loop_thread() = 0;
+
+  // Marshals a callback onto the message loop thread synchronously.
+  // The callback will be executed on the thread in FIFO order with all other
+  // tasks before this call returns.
+  //
+  // It's rarely a good idea to use blocking calls like this; if the loop is
+  // running large tasks (10-20ms/etc) the calling thread will block for that
+  // amount of time (or potentially longer if there's a large queue).
+  //
+  // This is safe to call from the loop thread (the callback will just be made
+  // inline), but that's probably not what you want.
+  virtual void MarshalSync(std::function<void()> callback) = 0;
+
+  // Marshals a callback onto the message loop thread asynchronously.
+  // The callback will be queued for execution on the thread in FIFO order with
+  // all other tasks and this call will return immediately.
+  //
+  // The task_list is used to safely cancel the callback and when the
+  // pending_task_list is destroyed all callbacks queued on it will be
+  // canceled. See TaskList for more information.
+  //
+  // Usage:
+  //  TaskList task_list;  // Can be kept as a class member, etc.
+  //  loop->MarshalAsync(&task_list, []() {
+  //    // Running on the loop thread.
+  //  });
+  virtual void MarshalAsync(TaskList* pending_task_list,
+                            std::function<void()> callback);
+
+  // Defers the function call until the next thread tick.
+  // This ensures that the callback will not run in the same call stack and can
+  // be treated as if it was asynchronous.
+  // The callback will not be executed if the loop exits before it is run or
+  // the TaskList is deallocated.
+  ref_ptr<Task> Defer(TaskList* pending_task_list,
+                      std::function<void()> callback);
+
+  // Defers the function call until the given delay elapses.
+  // The exact timing of the callback is undefined and it may occur earlier or
+  // later than the requested delay time.
+  // The callback will not be executed if the loop exits before it is run or
+  // the TaskList is deallocated.
+  ref_ptr<Task> Defer(TaskList* pending_task_list,
+                      std::function<void()> callback,
+                      std::chrono::milliseconds delay_millis);
+
+  // Defers the function call and repeatedly calls it until canceled.
+  // The exact timing of the callback is undefined and it may occur earlier or
```

Comment:
More guarantees here would be nice.
If the hosting loop ticks after 1 second, will a DeferRepeating(100ms) task run only once, until the next tick? I'm trying to compare this to something like Unity's Update() / FixedUpdate().